Assets/Scripts/CharacterHealth.cs

```diff
@@ -0,0 +1,42 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/// <summary>
+/// Parent class that holds basic features of HP
+/// </summary>
+abstract public class CharacterHealth : MonoBehaviour
+{
+    [Tooltip("Max health of this game character")]
+    public int MaxHealth;
+
+    [Tooltip("Current health points of the character")]
+    public int Health;
+
+    /// <summary>
+    /// Decrements the health and triggers death if appropriate
+    /// </summary>
+    /// <param name="damageValue">How much take to take</param>
+    public void TakeDamage(int damageValue)
+    {
+        Health -= damageValue;
+
+        if (Health <= 0) Die();
+    }
+
+    /// <summary>
+    /// Increments the health up to max health
+    /// </summary>
+    /// <param name="healValue"></param>
+    public void Heal(int healValue)
+    {
+        Health += healValue;
+
+        Health = Mathf.Max(Health, MaxHealth);
```

Comment:
I think this can overheal. Imagine the player is at 2 current HP, and has 100 total HP, Heal(1) would bring current HP to 100. I'd recommend removing line 33 and changing line 35 to "Health = Math.Min(Health + healValue, MaxHealth". Using Math instead of Mathf should also return an int, which would be good.

Reply:
Mathf returns an int due to overloading

Reply:
Oh, you're right! I was under the impression Mathf was all about floating point math, but now I think the f stands for functions.